GeoLib/OctTree-impl.h

```diff
@@ -0,0 +1,260 @@
+/**
+ * \date   2015-06-12
+ * \brief  Implementation of the OctTree class.
+ *
+ * \copyright
+ * Copyright (c) 2012-2015, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+
+namespace GeoLib {
+
+template <typename POINT>
+OctTree<POINT> OctTree<POINT>::createOctTree(POINT ll, POINT ur,
+	double eps, std::size_t max_items_per_node)
+{
+	OctTree<POINT>::_max_items_per_node = max_items_per_node;
+	OctTree<POINT>::_eps = eps;
+	// compute an axis aligned cube around the points ll and ur
+	const double dx(ur[0] - ll[0]);
+	const double dy(ur[1] - ll[1]);
+	const double dz(ur[2] - ll[2]);
+	if (dx >= dy && dx >= dz) {
+		ll[1] -= (dx-dy)/2.0;
+		ur[1] += (dx-dy)/2.0;
+		ll[2] -= (dx-dz)/2.0;
+		ur[2] += (dx-dz)/2.0;
+	} else {
+		if (dy >= dx && dy >= dz) {
+			ll[0] -= (dy-dx)/2.0;
+			ur[0] += (dy-dx)/2.0;
+			ll[2] -= (dy-dz)/2.0;
+			ur[2] += (dy-dz)/2.0;
+		} else {
+			ll[0] -= (dz-dx)/2.0;
+			ur[0] += (dz-dx)/2.0;
+			ll[1] -= (dz-dy)/2.0;
+			ur[1] += (dz-dy)/2.0;
+		}
+	}
+	return OctTree<POINT>(ll, ur);
+}
+
+template <typename POINT>
+OctTree<POINT>::~OctTree()
+{
+	for (std::size_t k(0); k < 8; ++k)
+		delete _childs[k];
+}
+
+template <typename POINT>
+bool OctTree<POINT>::addPoint(POINT * pnt, POINT *& ret_pnt)
```

Comment:
The return type could be a boost::optional<point> if not too difficult to change.
The input could be a reference to a point.